// zeptomail/lib/js/init.js
import fetch from 'node-fetch';

function _assertThisInitialized(e) {
  if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  return e;
}
function _callSuper(t, o, e) {
  return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e));
}
function _classCallCheck(a, n) {
  if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function");
}
function _defineProperties(e, r) {
  for (var t = 0; t < r.length; t++) {
    var o = r[t];
    o.enumerable = o.enumerable || false, o.configurable = true, "value" in o && (o.writable = true), Object.defineProperty(e, _toPropertyKey(o.key), o);
  }
}
function _createClass(e, r, t) {
  return r && _defineProperties(e.prototype, r), Object.defineProperty(e, "prototype", {
    writable: false
  }), e;
}
function _defineProperty(e, r, t) {
  return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, {
    value: t,
    enumerable: true,
    configurable: true,
    writable: true
  }) : e[r] = t, e;
}
function _getPrototypeOf(t) {
  return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) {
    return t.__proto__ || Object.getPrototypeOf(t);
  }, _getPrototypeOf(t);
}
function _inherits(t, e) {
  if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function");
  t.prototype = Object.create(e && e.prototype, {
    constructor: {
      value: t,
      writable: true,
      configurable: true
    }
  }), Object.defineProperty(t, "prototype", {
    writable: false
  }), e && _setPrototypeOf(t, e);
}
function _isNativeReflectConstruct() {
  try {
    var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {}));
  } catch (t) {}
  return (_isNativeReflectConstruct = function () {
    return !!t;
  })();
}
function ownKeys(e, r) {
  var t = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var o = Object.getOwnPropertySymbols(e);
    r && (o = o.filter(function (r) {
      return Object.getOwnPropertyDescriptor(e, r).enumerable;
    })), t.push.apply(t, o);
  }
  return t;
}
function _objectSpread2(e) {
  for (var r = 1; r < arguments.length; r++) {
    var t = null != arguments[r] ? arguments[r] : {};
    r % 2 ? ownKeys(Object(t), true).forEach(function (r) {
      _defineProperty(e, r, t[r]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) {
      Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
    });
  }
  return e;
}
function _possibleConstructorReturn(t, e) {
  if (e && ("object" == typeof e || "function" == typeof e)) return e;
  if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined");
  return _assertThisInitialized(t);
}
function _setPrototypeOf(t, e) {
  return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) {
    return t.__proto__ = e, t;
  }, _setPrototypeOf(t, e);
}
function _toPrimitive(t, r) {
  if ("object" != typeof t || !t) return t;
  var e = t[Symbol.toPrimitive];
  if (void 0 !== e) {
    var i = e.call(t, r);
    if ("object" != typeof i) return i;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (String )(t);
}
function _toPropertyKey(t) {
  var i = _toPrimitive(t, "string");
  return "symbol" == typeof i ? i : i + "";
}

var Client = /*#__PURE__*/function () {
  function Client(_ref, clientOptions) {
    var token = _ref.token,
      _ref$url = _ref.url,
      url = _ref$url === void 0 ? "" : _ref$url,
      debug = _ref.debug,
      _ref$domain = _ref.domain,
      domain = _ref$domain === void 0 ? "" : _ref$domain;
    _classCallCheck(this, Client);
    _defineProperty(this, "debug", void 0);
    _defineProperty(this, "commonURL", void 0);
    _defineProperty(this, "errorText", void 0);
    _defineProperty(this, "token", void 0);
    _defineProperty(this, "clientOption", void 0);
    _defineProperty(this, "isUrl", void 0);
    _defineProperty(this, "domain", void 0);
    _defineProperty(this, "defaultOption", void 0);
    this.debug = debug;
    this.commonURL = "api.zeptomail.".concat(domain || "com", "/");
    this.errorText = "";
    this.token = token;
    this.domain = domain;
    this.defaultOption = {
      host: url,
      isUrlAPI: url && url.indexOf("/v1.1") === -1,
      useHttps: !url.includes("https://"),
      authHeader: "Zoho-enczapikey"
    };
    this.clientOption = _objectSpread2(_objectSpread2({}, this.defaultOption), clientOptions);
    this.isUrl = url && url.indexOf("/v1.1") > -1;
  }
  // to get the appropriate correct URL
  return _createClass(Client, [{
    key: "getUrl",
    value: function getUrl(url) {
      var scheme = this.clientOption.useHttps ? "https://" : "";
      var host = this.isUrl ? this.clientOption.host : "".concat(this.clientOption.isUrlAPI ? this.clientOption.host : this.commonURL).concat(url);
      return "".concat(scheme).concat(host);
    }
    // to get header to request
  }, {
    key: "getHeader",
    value: function getHeader() {
      var headerValue = {
        'Authorization': this.token
      };
      return headerValue;
    }
    // check whether is it success or not
  }, {
    key: "status",
    value: function status(resp) {
      if (resp.ok) {
        return Promise.resolve(resp);
      }
      throw resp;
    }
    // convert Response to JSON
  }, {
    key: "toJSON",
    value: function toJSON(resp) {
      return resp.json();
    }
  }, {
    key: "validate",
    value: function validate(_ref2) {
      var options = _ref2.options;
      var valid = true;
      if (!this.token) {
        this.errorText = "Send Mail token cannot be empty";
      } else if (!this.clientOption.host && !this.domain) {
        this.errorText = "Provide a valid URL or a domain.";
      } else if (!options.to) {
        this.errorText = "To address cannot be empty.";
        valid = false;
      } else if (!options.from) {
        this.errorText = "From address cannot be empty.";
        valid = false;
      }
      return valid;
    }
  }, {
    key: "validateMail",
    value: function validateMail(_ref3) {
      var options = _ref3.options;
      var valid = this.validate({
        options: options
      });
      if (valid && !options.subject) {
        this.errorText = "Subject cannot be empty.";
        valid = false;
      }
      return valid;
    }
  }, {
    key: "validateTemplate",
    value: function validateTemplate(_ref4) {
      var options = _ref4.options;
      var valid = this.validate({
        options: options
      });
      if (valid && !options.mail_template_key && !options.template_alias && !options.template_key) {
        this.errorText = "Provide a valid template key.";
        valid = false;
      }
      return valid;
    }
  }, {
    key: "resultWithBody",
    value: function resultWithBody(method, query, url) {
      var options = query.options,
        isTemplate = query.isTemplate,
        instance = this,
        valid = true;
      if (isTemplate) {
        valid = this.validateTemplate({
          options: options
        });
      } else {
        valid = this.validateMail({
          options: options
        });
      }
      if (valid) {
        return this.httpRequest(fetch(instance.getUrl(url), {
          body: JSON.stringify(options),
          method: method,
          headers: this.getHeader()
        })).then(function (response) {
          return Promise.resolve(response);
        })["catch"](function (e) {
          return Promise.reject(e);
        });
      } else {
        return Promise.reject(instance.errorText);
      }
    }
  }, {
    key: "httpRequest",
    value: function httpRequest(fetched) {
      var resolve,
        reject,
        instance = this;
      var promise = new Promise(function (res, rej) {
        resolve = res;
        reject = rej;
      });
      fetched.then(instance.status).then(instance.toJSON).then(function (json) {
        /* to debug */
        if (instance.debug) {
          console.log(JSON.stringify("success"));
        }
        resolve(json);
      })["catch"](function (error) {
        if (instance.debug) {
          console.log(JSON.stringify("error"));
        }
        if (error.json) {
          instance.toJSON(error).then(function (error) {
            return reject(error);
          });
        } else {
          reject(error);
        }
      });
      return promise;
    }
  }]);
}();

var SendMailClient = /*#__PURE__*/function (_Client) {
  function SendMailClient(options) {
    var _this;
    var clientOption = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    _classCallCheck(this, SendMailClient);
    _this = _callSuper(this, SendMailClient, [options, clientOption]);
    _defineProperty(_this, "isUrl", void 0);
    var token = options.token,
      _options$url = options.url,
      url = _options$url === void 0 ? "" : _options$url,
      debug = options.debug;
    _this.token = token;
    _this.debug = debug;
    _this.isUrl = url && url.indexOf("/v1.1") > -1;
    return _this;
  }
  // to send mail
  _inherits(SendMailClient, _Client);
  return _createClass(SendMailClient, [{
    key: "sendMail",
    value: function sendMail(options) {
      /* to debug */
      if (this.debug) {
        console.log("send mail", this.token);
      }
      return this.resultWithBody("POST", {
        options: options
      }, "v1.1/email");
    }
    // to send batch mail
  }, {
    key: "sendBatchMail",
    value: function sendBatchMail(options) {
      if (this.debug) {
        console.log("send batch mail", this.token);
      }
      return this.resultWithBody("POST", {
        options: options
      }, "v1.1/email/batch");
    }
    // to send mail with template
  }, {
    key: "sendMailWithTemplate",
    value: function sendMailWithTemplate(options) {
      var query = {
        options: options,
        isTemplate: true
      };
      return this.resultWithBody("POST", query, "v1.1/email/template");
    }
  }, {
    key: "mailBatchWithTemplate",
    value: function mailBatchWithTemplate(options) {
      var query = {
        options: options,
        isTemplate: true
      };
      return this.resultWithBody("POST", query, "v1.1/email/template/batch");
    }
  }]);
}(Client);

export { SendMailClient };
